packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,446 @@
+
+        require(
+            fromAllocation[AIndex].destination == toAllocation[AIndex].destination &&
+                fromAllocation[BIndex].destination == toAllocation[BIndex].destination &&
+                fromAllocation[IIndex].destination == toAllocation[IIndex].destination,
+            'destinations may not change'
+        );
+
+        require(fromAllocation[IIndex].amount == toAllocation[IIndex].amount, 'p2.amt !constant');
+        require(
+            fromAllocation[AIndex].amount +
+                fromAllocation[BIndex].amount +
+                fromAllocation[IIndex].amount ==
+                toAllocation[AIndex].amount +
+                    toAllocation[BIndex].amount +
+                    toAllocation[IIndex].amount,
+            'total allocation changed'
+        );
+
+        // Allowed named-state transitions are
+        //    AB
+        //    ^^
+        //   /  \
+        // A      B
+        // ^      ^
+        //  \    /
+        //   None
+
+        if (fromAppData.alreadyMoved == AlreadyMoved.None) {
+            require(
+                (toAppData.alreadyMoved == AlreadyMoved.A && signedByA) ||
+                    (toAppData.alreadyMoved == AlreadyMoved.B && signedByB),
+                'incorrect move from None'
+            );
+        } else {
+            // If a support proof has already been supplied, the current support proof must be greater
+            require(
+                toAppData.supportProofForX.turnNumTo > fromAppData.supportProofForX.turnNumTo,
+                'inferior support proof'
+            );
+            if (fromAppData.alreadyMoved == AlreadyMoved.A) {
+                require(
+                    toAppData.alreadyMoved == AlreadyMoved.AB && signedByB,
+                    'incorrect move from A'
+                );
+            } else if (fromAppData.alreadyMoved == AlreadyMoved.B) {
+                require(
+                    toAppData.alreadyMoved == AlreadyMoved.AB && signedByA,
+                    'incorrect move from B'
+                );
+            } else {
+                revert('move from None,A,B only');
+            }
+        }
+
+        // validate the supplied support proof
+        // extract the allocation
+        Outcome.AllocationItem[] memory Xallocation = validateSupportProofForX(
+            fromAppData,
+            toAppData
+        );
+
+        // ensure A,B part of the outcome of X has been absorbed into the outcome of J
+        require(
+            Xallocation[AIndex].amount == toAllocation[AIndex].amount &&
+                Xallocation[BIndex].amount == toAllocation[BIndex].amount &&
+                Xallocation[AIndex].destination == toAllocation[AIndex].destination &&
+                Xallocation[BIndex].destination == toAllocation[BIndex].destination,
+            'X / J outcome mismatch'
+        );
+        return true;
+    }
+
+    function validateSupportProofForX(AppData memory fromAppData, AppData memory toAppData)
```

Comment:
On discussion with @kerzhner I decided _not_ to dig into the implementation of this. We should merge this, and treat it as "prototype/reference code", and _not_ as code that we believe to be safe enough to use in the virtual funding protocol.